### DISCUSSION TITLE
AssertionScope w/explicit name AND DetermineCallerIdentity

### DISCUSSION THREAD

**Discussion by btzdnl** [fluentassertions/fluentassertions#2633] [Ideas]
Sometimes, I'd like to provide human-readable context for a particular block of assertions. This however means that FluentAssertions no longer does its automatic assertion subject check. So instead of...

> Expected instruction.Hidden not to be False, but found False.

I then get

> Expected Instruction 3 not to be False, but found False.

That's not terribly helpful in my opinion. I can no longer see the property being checked. I'd like to have both: Explicitly provided scope name AND the automatic subject detection. Or maybe even, generally, a way for the resulting context name to be made up of all the nested (named) scopes along the way.

I see that `AssertionScope` is currently being redesigned, so maybe this here is a possible future improvement?

**Comment by dennisdoomen** [comment]
Isn't that what https://github.com/fluentassertions/fluentassertions/pull/2607 fixes?

**Reply by btzdnl** [reply]
Unfortunately not, the “problematic” behavior is unchanged. As soon as one explicitly named context is in play, `DetermineCallerIdentity` is no longer used.

```cs
using var scope1 = new AssertionScope("Test1");
using var scope2 = new AssertionScope("Test2");
nonEmptyList.Should().BeEmpty();
```

means I get:

> Expected Test1/Test2 to be empty, but found at least one item {"Hello"}.

Previously, “Test1” wouldn’t have been mentioned, but even now I still don’t see what exactly what checked inside that scope.

**Reply by dennisdoomen** [reply]
Ah, you are expecting something like `Did not expect test1/tests/nonEmptyList to be empty"?

**Reply by btzdnl** [reply]
Yes, exactly!

**Reply by parched** [reply]
> Ah, you are expecting something like `Did not expect test1/tests/nonEmptyList to be empty"?

That's what I was expecting too, seems the most intuitive to me.

What's the best work around for this currently?

**Reply by dennisdoomen** [reply]
This is how it will work for 8.0 (to be released today)... https://github.com/fluentassertions/fluentassertions/blob/f77fe325fb03040d8b32973a0377ef9213990ba4/Tests/FluentAssertions.Specs/Execution/AssertionScopeSpecs.cs#L310
